distributed/deploy/subprocess.py

```diff
@@ -15,13 +16,90 @@
 from distributed.compatibility import WINDOWS
 from distributed.deploy.spec import ProcessInterface, SpecCluster
 from distributed.deploy.utils import nprocesses_nthreads
-from distributed.scheduler import Scheduler
 from distributed.worker_memory import parse_memory_limit
 
 logger = logging.getLogger(__name__)
 
 
-class SubprocessWorker(ProcessInterface):
+class Subprocess(ProcessInterface, abc.ABC):
+    process: asyncio.subprocess.Process | None
+
+    def __init__(self):
+        if WINDOWS:
+            # FIXME: distributed#7434
+            raise RuntimeError("Subprocess does not support Windows.")
+        self.process = None
+        super().__init__()
+
+    async def start(self) -> None:
+        await self._start()
+        await super().start()
+
+    @abc.abstractmethod
+    async def _start(self) -> None:
+        """Start the subprocess"""
+
+    async def close(self) -> None:
+        if self.process and self.process.returncode is None:
+            for child in psutil.Process(self.process.pid).children(recursive=True):
+                child.kill()
+            self.process.kill()
+            await self.process.communicate()
+        self.process = None
+        await super().close()
+
+
+class SubprocessScheduler(Subprocess):
+    """A local Dask scheduler running in a dedicated subprocess
+
+    Parameters
+    ----------
+    scheduler_kwargs:
+        Keywords to pass on to the ``Scheduler`` class constructor
+    """
+
+    scheduler_kwargs: dict
+    address: str | None
+
+    def __init__(
+        self,
+        scheduler_kwargs: dict | None = None,
+    ):
+        self.scheduler_kwargs = scheduler_kwargs or {}
+        super().__init__()
+
+    async def _start(self):
+        cmd = [
+            "dask",
+            "spec",
+            "--spec",
+            json.dumps(
+                {"cls": "distributed.Scheduler", "opts": {**self.scheduler_kwargs}}
+            ),
+        ]
+        logger.info(" ".join(cmd))
+        self.process = await asyncio.create_subprocess_exec(
+            "dask",
+            "spec",
+            "--spec",
+            json.dumps(
+                {"cls": "distributed.Scheduler", "opts": {**self.scheduler_kwargs}}
+            ),
+            stderr=asyncio.subprocess.PIPE,
+        )
+
+        while True:
+            line = (await self.process.stderr.readline()).decode()
+            if not line.strip():
+                raise RuntimeError("Scheduler failed to start")
+            logger.info(line.strip())
```

Comment:
```suggestion
            sys.stderr.write(line)
```
We're not redirecting stderr from the workers; seems like output from the scheduler should be treated the same? Subtle, but logging could be configured differently from plain stderr—including to prefix some additional information which could be confusing—so just forwarding to stderr seems more consistent with the worker subprocesses.

Reply:
See https://github.com/dask/distributed/pull/7727#discussion_r1158853815 for the general discrepancy between the generally configured log level and the one required to retrieve the address.